### PR TITLE
fix: screenchar()/screenstring() with hidden floating windows

### LINKED ISSUE #35556: `screenstring()` and `screenattr()` do not respect hidden floating windows
### Problem

Having extui/ui2 enabled makes `screenstring()` and `screenattr()` return not what is actually on screen.

Cc @luukvbaal 

### Steps to reproduce

1. Create 'init.lua' with the following content:

    ```lua
    _G.get_screenshot = function()
      local s, a = {}, {}
      for i = 1, 6 do
        s[i], a[i] = {}, {}
        for j = 1, 10 do
          s[i][j] = vim.fn.screenstring(i, j)
          a[i][j] = vim.fn.screenattr(i, j)
        end
      end

      return s, a
    end

    vim.o.winborder = 'double'
    local win_opts = { relative = 'editor', row = 0, col = 0, width = 5, height = 1, hide = true }
    vim.api.nvim_open_win(vim.api.nvim_create_buf(false, true), false, win_opts)

    -- Prepare buffer text to easier distinguish between floating and normal window
    local x_line = string.rep('x', 20)
    local lines = { x_line, x_line, x_line, x_line, x_line }
    vim.api.nvim_buf_set_lines(0, 0, -1, false, lines)
    ```
2. `nvim --clean -u init.lua`
3. `:=get_screenshot()`. The output is as follows (formatting is mine for clarity):
    ```
    {
      { "╔", "═", "═", "═", "═", "═", "╗", "x", "x", "x" },
      { "║", " ", " ", " ", " ", " ", "║", "x", "x", "x" },
      { "╚", "═", "═", "═", "═", "═", "╝", "x", "x", "x" },
      { "x", "x", "x", "x", "x","x", "x", "x", "x", "x" },
      { "x", "x", "x", "x", "x", "x", "x", "x", "x", "x" },
      { "~", " ", " ", " ", " ", " ", " ", " ", " ", " " }
    }
    {
      { 30, 30, 30, 30, 30, 30, 30, 0, 0, 0 },
      { 30, 30, 30, 30, 30, 30, 30, 0, 0, 0 },
      { 30, 30, 30, 30, 30, 30, 30, 0, 0, 0 },
      { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 },
      { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 },
      { 18, 18, 18, 18, 18, 18, 18, 18, 18, 18 }
    }
    ```

Not setting 'winborder' still has a single top line show not expected results.

Might be related to #34295.

### Expected behavior

Both `screenstring()` and `screenattr()` always show what is visible on screen.

### Nvim version (nvim -v)

NVIM v0.12.0-dev-1113+g4edeaaa6e2

### Vim (not Nvim) behaves the same?

No, doesn't have this functionality

### Operating system/version

EndeavourOS Linux x86_64, 6.16.3-arch1-1

### Terminal name/version

Ghostty 1.1.3-arch1

### $TERM environment variable

xterm-ghostty

### Installation

Appimage

### PR DESCRIPTION
**Pull request by echasnovski**
Fix #35556 .

---

I need a bit of guidance here. The added `from ui2` test passes both before and with this PR, but the same setup when tested manually with `nvim --clean` does not work before PR and works after PR. This doesn't happen with `'manual'` test, which behaves as expected (fails before PR and passes with PR).

How can I adjust the `from ui2` test to fail before this PR?

### CODE REVIEW

File: test/functional/vimscript/screenchar_spec.lua [hunk 1]
```diff
@@ -79,4 +79,34 @@ describe('screenchar() and family respect floating windows', function()
...
   end)
+
+  describe('hidden windows', function()
+    before_each(function()
+      clear()
+      Screen.new(40, 7, {})
+      api.nvim_buf_set_lines(0, 0, -1, true, { 'aaa', 'aaa' })
+    end)
+
+    local validate = function()
```
Comment (justinmk) [inline test/functional/vimscript/screenchar_spec.lua]: let's call this `assert_something`, since that's the common convention here, and `validate` is very well established for `vim.validate`. also we always use `local function` where possible.
```suggestion
    local function assert_...()
```

**Comment by justinmk** [comment]
> How can I adjust the `from ui2` test to fail before this PR?

maybe @luukvbaal has insight. 

doesn't need to block this though, if it fixes an issue, can be a followup.

**Comment by luukvbaal** [comment]
The ui2 windows are currently only created upon the first event, so adding `command('echo "foo"')` should make it fail. Though it seems unnecessary to add a ui2 test here as the "manual" test tests the same code path, while being more explicit.

**Comment by echasnovski** [comment]
> The ui2 windows are currently only created upon the first event, so adding `command('echo "foo"')` should make it fail.

Yes, this does seem to make it fail before the PR and pass after. Thanks! I'll push tomorrow.

> Though it seems unnecessary to add a ui2 test here as the "manual" test tests the same code path, while being more explicit.

I did think about that, but my main use case is for `screenstring()` family of functions to work if/when ui2 is enabled by default. Maybe it will use something custom in the future? One explicit small test should not be a big problem, while testing exactly the core issue.

**Comment by echasnovski** [comment]
Adjusted tests. Note for possible future use: I tried to just add another hidden window inside `setup_floating_windows()` for both more compact and covering tests, but it didn't fail before this PR. It looks like the failing edge case here is when there are no visible floating windows at target cell.
